src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -208,12 +195,16 @@ SyntaxKind.TupleExpression or
                     return false;
                 }
 
+                // We know the current node has delimiters due to the Kind() check above, so isInsideDelimiters is
+                // simply the inverse of being outside the delimiters.
+                isInsideDelimiters = !HasDelimitersButCaretIsOutside(currentNode, caret.Position);
+
                 var newCaret = args.SubjectBuffer.CurrentSnapshot.GetPoint(newCaretPosition);
                 if (!TryGetStartingNode(root, newCaret, out currentNode, cancellationToken))
                     return false;
 
                 return MoveCaretToSemicolonPosition(
-                    speculative, args, document, root, originalCaret, newCaret, syntaxFacts, currentNode, isInsideDelimiters: true, cancellationToken);
```

Comment:
📝 The root cause of the bug is assuming `isInsideDelimiters` was true here. Normally it was true due to the way `TryGetStartingNode` skips nodes where we are outside delimiters, but `MoveCaretToSemicolonPosition` is called recursively which allowed us to violate this expectation.